jmclient/jmclient/payjoin.py

```diff
@@ -0,0 +1,503 @@
+        # check if the hostname is in the the whitelist,
+        # otherwise return the default policy
+        if hostname in self.good_domains:
+            return CertificateOptions(verify=False)
+        return self.default_policy.creatorForNetloc(hostname, port)
+
+class JMPayjoinManager(object):
+    """ An encapsulation of state for an
+    ongoing Payjoin payment. Allows reporting
+    details of the outcome of a Payjoin attempt.
+    """
+
+    # enum such that progress can be
+    # reported
+    JM_PJ_NONE = 0
+    JM_PJ_INIT = 1
+    JM_PJ_PAYMENT_CREATED = 2
+    JM_PJ_PAYMENT_SENT = 3
+    JM_PJ_PARTIAL_RECEIVED = 4
+    JM_PJ_PARTIAL_REJECTED = 5
+    JM_PJ_PAYJOIN_COSIGNED = 6
+    JM_PJ_PAYJOIN_BROADCAST = 7
+    JM_PJ_PAYJOIN_BROADCAST_FAILED = 8
+
+    pj_state = JM_PJ_NONE
+
+    def __init__(self, wallet_service, mixdepth, destination,
+                 amount, server):
+        assert isinstance(wallet_service, WalletService)
+        # payjoin is not supported for non-segwit wallets:
+        assert isinstance(wallet_service.wallet,
+                          (SegwitWallet, SegwitLegacyWallet))
+        # our payjoin implementation requires PSBT
+        assert isinstance(wallet_service.wallet, PSBTWalletMixin)
+        self.wallet_service = wallet_service
+        # mixdepth from which payment is sourced
+        assert isinstance(mixdepth, int)
+        self.mixdepth = mixdepth
+        assert isinstance(destination, btc.CCoinAddress)
+        self.destination = destination
+        assert isinstance(amount, int)
+        assert amount > 0
+        self.amount = amount
+        self.server = server
+        self.pj_state = self.JM_PJ_INIT
+        self.payment_tx = None
+        self.initial_psbt = None
+        self.payjoin_psbt = None
+        self.final_psbt = None
+        # change is initialized as None
+        # in case there is no change:
+        self.change_out = None
+
+    def set_payment_tx_and_psbt(self, in_psbt):
+        assert isinstance(in_psbt, btc.PartiallySignedTransaction)
+        self.initial_psbt = in_psbt
+        # any failure here is a coding error, as it is fully
+        # under our control.
+        assert self.sanity_check_initial_payment()
+        self.pj_state = self.JM_PJ_PAYMENT_CREATED
+
+    def sanity_check_initial_payment(self):
+        """ These checks are motivated by the checks specified
+        for the *receiver* in the btcpayserver implementation doc.
+        We want to make sure our payment isn't rejected.
+        We also sanity check that the payment details match
+        the initialization of this Manager object.
+        """
+        # failure to extract tx should throw an error;
+        # this PSBT must be finalized and sane.
+        self.payment_tx = self.initial_psbt.extract_transaction()
+
+        # inputs must all have witness utxo populated
+        for inp in self.initial_psbt.inputs:
+            if not inp.utxo and isinstance(inp.utxo, btc.CTxOut):
```

Comment:
This check won't work in general after this PR to python-bitcointx will be merged: https://github.com/Simplexum/python-bitcointx/pull/35 

That PR will allow to supply `CTransaction` as utxo to witness input (because there's known attack with signing two times with different segwit inputs, and workaround is supplying full transaction in psbt input even for segwit). So `inp.utxo` may be `CTransaction` or `CTxOut` depending on what was supplied at `PSBT_Input` creation, and additional `inp.witness_utxo` and `inp.non_witness_utxo` will be either `None` or `CTxOut`, `CTransaction` respectively.

Reply:
Fixed to reference `inp.witness_utxo` in  9933408ad70e717014ecbb24796d104f3263ac05